rl.py

```diff
@@ -0,0 +1,217 @@
+"""
+This script trains a minimal policy gradient agent on the "Toy environment", which is the
+defined in Figure 2 of the GFlowNet Foundations paper, Bengio et al (JMLR, 2023):
+
+.. _a link: https://jmlr.org/papers/v24/22-0364.html
+"""
+
+"""
+Key differences with gfn.py:
+- Uses policy gradient instead of GFlowNet training
+- Batches of episodes instead of single episodes to reduce variance
+"""
+
+import torch
+from torch.distributions import Categorical
+from tqdm import tqdm
+import time
+
+
+### COMMON VARIABLES ###
+
+float_type = torch.float32
+device = torch.device("cpu")
+do_print = True
+
+### ENVIRONMENT ###
+
+discount_factor = 1.0 # No discounting
+batch_size = 32
+
+# A dictionary of connections: the keys of the dictionary are the indices of the
+# states, and the values are the indices of the states to which each state is
+# connected.
+connections_dict = {
+    0: (1, 2),
+    1: (3,),
+    2: (3, 4),
+    3: (5, -1),
+    4: (6, -1),
+    5: (7, 8),
+    6: (8, 10, -1),
+    7: (9,),
+    8: (9, -1),
+    9: (-1,),
+    10: (-1,),
+}
+# A dictionary of rewards: the keys of the dictionary are the indices of the
+# states, and the values are their rewards.
+rewards_dict = {
+    3: 30,
+    4: 14,
+    6: 23,
+    8: 10,
+    9: 30,
+    10: 5,
+}
+n_states = len(connections_dict)
+
+### POLICY MODEL ###
+
+# Linear layer following an embedding layer. The inputs are the state indices and the
+# outputs are tensors with dimensionality equal to the number of actions: the number of
+# states plus one, for the end-of-sequence (EOS) action.
+class Policy(torch.nn.Module):
+    def __init__(self, n_states, float_type, device):
+        super(Policy, self).__init__()
+        self.embedding = torch.nn.Embedding(n_states, n_states)
+        self.linear = torch.nn.Linear(n_states, n_states + 1, dtype=float_type, device=device)
+
+    def forward(self, x):
+        x = self.embedding(x)
+        x = self.linear(x)
+        return x
+
+policy = Policy(n_states, float_type, device)
+
+### OPTIMIZER ###
+
+n_train_steps = 2000
+learning_rate = 0.01
+optimizer = torch.optim.Adam(policy.parameters(), lr=learning_rate)
```

Comment:
I would use the same optimizer (SGD with momentum) as in the GFN implementation, for more direct comparison.